library/core/src/primitive_docs.rs

```diff
@@ -1506,6 +1506,110 @@ mod prim_ref {}
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else calling the function
+/// via that function pointer is Undefined Behavior. ABI compatibility is a lot stricter than merely
+/// having the same memory layout; for example, even if `i32` and `f32` have the same size and
+/// alignment, they might be passed in different registers and hence not be ABI-compatible.
+///
+/// ABI compatibility as a concern only arises in code that alters the type of function pointers,
+/// and in code that combines `#[target_feature]` with `extern fn`. Altering the type of
+/// function pointers is wildly unsafe (as in, a lot more unsafe than even
+/// [`transmute_copy`][mem::transmute_copy]), and should only occur in the most exceptional
+/// circumstances. `#[target_feature]` is also used rarely. But assuming such circumstances, what
+/// are the rules?
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible ABI string,
+/// must take the same number of arguments, the individual argument types and the return types must
+/// be ABI-compatible, and the target feature requirements must be met (see the subsection below for
+/// the last point). The ABI string is declared via `extern "ABI" fn(...) -> ...`; note that
+/// `fn name(...) -> ...` implicitly uses the `"Rust"` ABI string and `extern fn name(...) -> ...`
+/// implicitly uses the `"C"` ABI string.
+///
+/// The ABI strings are guaranteed to be compatible if they are the same, or if the caller ABI
+/// string is `$X-unwind` and the callee ABI string is `$X`, where `$X` is one of the following:
+/// "C", "aapcs", "fastcall", "stdcall", "system", "sysv64", "thiscall", "vectorcall", "win64".
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>` (specifically, only `Box<T, Global>`),
+///   `NonNull<T>` are all ABI-compatible with each other for all `T`. Two of these pointer types
+///   with different `T` are ABI-compatible if they have the same metadata type (`<T as
+///   Pointee>::Metadata`).
+/// - `usize` is ABI-compatible with the `uN` integer type of the same size, and likewise `isize` is
+///   ABI-compatible with the `iN` integer type of the same size.
+/// - Any two `fn` types are ABI-compatible with each other if they have the same ABI string or the
+///   ABI string only differs in a trailing `-unwind`, independent of the rest of their signature.
+///   (Note that this is about the case of passing a function pointer as an argument to a function.
+///   The two pointers being ABI-compatible here means that the call successfully passes the
+///   pointer. When actually calling the pointer, of course the rest of the signature becomes
```

Comment:
I am not sure what you mean by "successfully passes the pointer." "calling the pointer" doesn't make sense to me either. Did you mean "calling the function through the pointer"?

Reply:
> I am not sure what you mean by "successfully passes the pointer." 

There's no UB from passing the function pointer around.

IOW, if the caller has signature `fn(fn())`, and the callee has signature `fn(fn(i32) -> i32)`, then the function call itself is completely well-defined. Caller and callee are ABI compatible even though the only argument of this function has a different type in the two signatures.

Of course if the callee actually calls the function pointer that it was given as argument, then the fact that the signatures are different matters.

> "calling the pointer" doesn't make sense to me either. 

When `f: fn()`, then `f()` is "calling the function pointer". At least that's how I would call that operation. You seem to call it "calling the function through the pointer".

With data pointers we say that we read and write the pointer, we don't always spell out "read and write the memory through the pointer". I am following the same principle for function pointers.

Reply:
This point here is a bit confusing since the question "are two fn ptr types A and B ABI-compatible" is ambiguous:
- it could mean "can I call a function of type A with a caller-side signature of type B"
- it could mean "can I call a function of type `fn(A)` with a caller-side signature of type `fn(B)`"

Usually when we say, e.g., "`u32` and `NonZeroU32` are ABI-compatible", we mean the latter, but when the two types in questions are themselves function pointers then the terminology becomes unclear.

I'd be happy for suggestions for how to word this more clearly.

Reply:
> This point here is a bit confusing since the question "are two fn ptr types A and B ABI-compatible" is ambiguous:
>
> it could mean "can I call a function of type A with a caller-side signature of type B"
> it could mean "can I call a function of type fn(A) with a caller-side signature of type fn(B)"
>
> Usually when we say, e.g., "u32 and NonZeroU32 are ABI-compatible", we mean the latter, but when
> the two types in questions are themselves function pointers then the terminology becomes unclear.
>
> I'd be happy for suggestions for how to word this more clearly.

First, I admit I don't think it is good to guarantee anything about transmuting function pointers with incompatible signatures, as I don't understand the motivation driving us to make such guarantees. So one way to clarify things would be to just have one set of rules for ABI compatibility / transmutation for function types for now, and then maybe follow up later with a proposal to guarantee transmutations of incompatible function types work.

> usually when we say, e.g., "u32 and NonZeroU32 are ABI-compatible", 

If you reserve "ABI compatibility" to be strictly about function calls, function declarations, function definitions, and function pointers, then we could have a separate term "transmutable, e.g. "NonZeroU32 is transmutable u32" (where "transmutable" means the result of the transmutation is well-defined), and we could have a separate term "A is argument-compatible with B" to talk about where an argument of type A can be passed for a function parameter of type B. And hopefully we would have rules like "Any type A that is transmutable to B is argument-compatible with B." Then you could use "argument-compatible" as part of the definition of ABI-compatibility, in particular each argument in a function-call must be argument-compatible with corresponding parameter in the function's definition (and later, declaration), and (target_feature stuff, etc.).

In other words, don't use "ABI-compatible" and "ABI compatibility" terms to talk about argument/parameter compatibility.

Reply:
> First, I admit I don't think it is good to guarantee anything about transmuting function pointers with incompatible signatures, as I don't understand the motivation driving us to make such guarantees. So one way to clarify things would be to just have one set of rules for ABI compatibility / transmutation for function types for now, and then maybe follow up later with a proposal to guarantee transmutations of incompatible function types work.

I think it would be really strange to say that `*const i32` and `*mut u8` are ABI-compatible, but `fn()` and `fn(i32)` are not. Both are just pointer types with different pointee information. It'd be very surprising to require the function signatures to be ABI compatible in order to have the pointers be ABI compatible.

> transmutable

That's a very bad term for this situation. `i32` is transmutable to `u32` but they are not ABI compatible. The PR explicitly talks about this.

For better or worse, "ABI compatible" is already in common use for this concept, in questions like "are `u32` and `char` ABI compaible?" I think it's not a bad term, it's only a bit annoying for this specific case of ABI compatibility of function pointer types.

I have updated the text to use an example instead of remaining so abstract, I hope that helps.